File: src/org/newtonpolyhedron/solverprinters/ConeSolverPrinter.java
package org.newtonpolyhedron.solverprinters;

import static java.text.MessageFormat.*;

import java.io.PrintWriter;
import java.util.List;

import org.newtonpolyhedron.entity.SolverPrinter;
import org.newtonpolyhedron.entity.vector.IntVector;
import org.newtonpolyhedron.solve.cone.ConeSolver;
import org.newtonpolyhedron.utils.MatrixUtils;

public class ConeSolverPrinter extends SolverPrinter <ConeSolver> {
	
	private final List <IntVector>	inequations;
	private final List <IntVector>	basis;
	
	public ConeSolverPrinter(
			final ConeSolver solver,
			final List <IntVector> inequations,
			final List <IntVector> basis,
			final PrintWriter output) {
		super(solver, output);
		this.inequations = inequations;
		this.basis = basis;
	}
	
	@Override
	protected void solveFor(final ConeSolver solver, final PrintWriter output) throws Exception {
		output.println(title("Cone computing"));
		final int rank = MatrixUtils.getRank(MatrixUtils.fromIntVector(inequations));
		output.println("Matrix rank = " + rank);
		output.println(header("Original inequalities:"));
		for (int i = 0; i < inequations.size(); i++) {
			output.println(format(" c{0} = {1}", i + 1, inequations.get(i)));
		}
		final int dim = inequations.get(0).getDim();
		final List <IntVector> solved = solver.solve(inequations, basis, dim, output);
		coneFinalSolutionOutput(solved, output);
	}
	
	private static void coneFinalSolutionOutput(
			final List <IntVector> testing,
			final PrintWriter output) {
		output.println(header("FINAL SOLUTIONS:"));
		for (int i = 0; i < testing.size(); i++) {
			final String str = testing.get(i).toString();
			final int len = str.length();
			output.print(str + "\t");
			if (len < 10) {
				output.print("\t");
			}
			if (len < 16) {
				output.print("\t");
			}
			if ((i + 1) % 4 == 0) {
				output.println();
			}
		}
		output.println(header("ANSWERS:"));
		for (final IntVector p : testing) {
			output.println(p);
		}
	}
}

File: src/org/newtonpolyhedron/solverprinters/ConeSolverPrinter.scala
package org.newtonpolyhedron.solverprinters

import java.io.PrintWriter
import java.text.MessageFormat

import org.newtonpolyhedron._
import org.newtonpolyhedron.entity.MatrixSupport
import org.newtonpolyhedron.entity.SolverPrinter
import org.newtonpolyhedron.entity.vector.VectorImports._
import org.newtonpolyhedron.solve.cone.ConeSolver

class ConeSolverPrinter(solver: ConeSolver,
                        val inequations: IndexedSeq[IntVec],
                        val basis: IndexedSeq[IntVec],
                        output: PrintWriter)
    extends SolverPrinter[ConeSolver](solver, output) {

  override def solveFor(solver: ConeSolver,
                        output: PrintWriter) = {
    output.println(title("Cone computing"))
    val rank = MatrixSupport.fromInts(inequations).rank
    output.println("Matrix rank = " + rank)
    output.println(header("Original inequalities:"))
    inequations eachWithIndex { (currIneq, i) =>
      output.println(MessageFormat.format(" c{0} = {1}", int2Integer(i + 1), currIneq))
    }
    val dimension = inequations(0).size
    val solved = solver.solve(inequations, basis, dimension, output)
    coneFinalSolutionOutput(solved, output)
  }

  private def coneFinalSolutionOutput(testing: Seq[IntVec],
                                      output: PrintWriter) = {
    output.println(header("FINAL SOLUTIONS:"))
    testing eachWithIndex { (currTesting, i) =>
      val str = currTesting.toString
      val len = str.length
      output.print(str + "\t")
      if (len < 10) {
        output.print("\t")
      }
      if (len < 16) {
        output.print("\t")
      }
      if ((i + 1) % 4 == 0) {
        output.println()
      }
    }
    output.println(header("ANSWERS:"))
    testing foreach (output.println)
  }
}